### PR TITLE
added new types. Need to work on parsing and language.

### PR DESCRIPTION
**Pull request by NCrusher74**
Sorry, it took me a while to get the Git wrinkles worked out. Since my branch was based off my forked master, it was a mess.

Anyway, this is what I have so far. Mostly I'm stuck as what I need to be doing in terms of parsing, particularly for the language and description strings, since the description string needs a string terminator and the language string shouldn't be visible in the frame content.

### CODE REVIEW

**Comment by NCrusher74** [comment]
Also, I intend to work on the CHAP and CTOC frames after I get this worked out, so we may just want to work on those while we're at it with these? I can start on the basics, probably, but those are an type of frame with potential to use subframes and it will be a lot trickier.

**Comment by NCrusher74** [comment]
I've done some more work on this; I think I may have worked out the stringToBytes conversion for the new frame types, and maybe even the string terminator for the description string. But I'm having a hard time with the language code string conformity. I'm not sure how to initialize the enum so that it will satisfy all the differnent ways it's being called.

**Comment by NCrusher74** [comment]
Okay, I think I've gotten about as far as I can with this. The frame creation tests are passing, as are a couple others.

There's a test for the CommentTypes adapter that isn't passing, but that's because I don't know how to translate the test content into bytes.

The acceptance tests are failing, but that is because the generated files won't write (therefore I can't add them to the bundle to test against) and I can't figure out why.

**Comment by NCrusher74** [comment]
Well, I got the writing issue worked out, I think. At least, I was able to write the new files. But there's an issue reading them. The reading part of the acceptance tests gives me an error:

```
caught "NSRangeException", "*** -[Foundation.__NSSwiftData subdataWithRange:]: range {457, 7895046} exceeds data length 158702"
```

I suspect maybe this is a problem with messing with the `bytesToString` adaptation stuff? I don't know how to fix it.

Also, I checked the written files in several other metadata readers (Yate, Kid3, and MediaInfo) as well as Itunes and Fission (an audio editor) and the contents of the comments and lyrics frame aren't showing up. Kid3 shows me the following:

Screen Shot 2020-03-13 at 12 29 54 PM

So it looks like my language code isn't getting adapted to a string properly.

It's also showing me this:

Screen Shot 2020-03-13 at 12 30 06 PM

So while we're here, we might want to fix the copyright frame.

**Comment by NCrusher74** [comment]
I've still got a number of tests failing, but I tend to work on the weekend a lot so I might not be able to try to fix them for a few days. But since the `UserTextInformation` frame needs a terminated `description` string to be compliant, I added a new frame type with accompanying parsing and adapting to account for it. Basically, it's the lyrics/comments frame without the language.

**Comment by NCrusher74** [comment]
In an effort to try to track down what is causing the fatalError with your PathBuilder on line 13, I imported this branch of the package to my other project I'm working on and ran the acceptance tests there. It failed with the following, highly unhelpful error:

```
caught "NSRangeException", "*** -[Foundation.__NSSwiftData subdataWithRange:]: range {457, 7234566} exceeds data length 158663"
```

So it looks like something--probably one of my string parsers or adapters--isn't playing well with either `Mp3WithID3TagBuilder` or one of your string or frame parsers.

I have no idea how to correct for that. I was mostly guessing as it was. Hopefully we can track this down.

**Comment by chicio** [comment]
Today I don't have time to watch but one thing I already noticed, not related your error, is that the code quality is decreasing too much. We are already under 3.90, but my focus is to keep it above 3.95. Check the codebeat report https://codebeat.co/projects/github-com-chicio-id3tageditor-master/pull_requests/323104

**Comment by NCrusher74** [comment]
I noticed that, but 8 of the 23 issues are related to the ISO-639-2 codes enum. I could get rid of the `languageFullToCode` and `languageCodeToFull` extensions, which would get rid of about six of those issues, but that would limit the options for the users as far as how they want to implement the language codes in their apps. Actually I probably will get rid of `languageCodeToFull` because I don't think having a variable that translates the code and returns the full language name will be useful to anyone.

Beyond that, there's nothing I can do about the length of the function, and I don't even know what "cyclomatic complexity" means.

Several more of the issues pertain to functions with too many arguments. Unfortunately, there's nothing I can do about that. Some of these frames we're implementing are going to have a lot of parameters (CTOC and CHAP each have six, and that doesn't include the frame identifier and version parameters for the `createFrame` functions.) 

Other places, what is being flagged is the code being too similar in two or more places, A lot of this is me following your pattern of giving each frame its own Frame Creator and Frame Content Parsing Operation Factory type.

Long story short? I'm only going to be able to fix a handful of those issues. Either they're included because I'm following your model, or they're relating to something I have no way of changing.

If you'd like to offer me guidance on how to better refine the code, I'd be grateful.

**Comment by NCrusher74** [comment]
Well, I made some changes to how the language code <-> language conversions are happening and we're at 3.96, so hopefully that will hold it for now.

**Comment by NCrusher74** [comment]
I'm trying to think of the best way to handle the the Content Parsing and Adapters for the CTOC and CHAP frames. It's more complicated than anything I've done so far, because both of them can have subframes. For instance, here is what I have so far for `ID3TableOfContentsStringAdapter`:

``swift
class ID3TableOfContentsStringAdapter: TableOfContentsStringAdapter {
     
    func adapt(content: String) -> ID3FrameTableOfContents {
        let contentComponents = content.components(separatedBy: ":")
        let elementID = getElementIDFrom(contentComponents: contentComponents)
        let topLevelFlag = getTopLevelFlagFrom(contentComponents: contentComponents)
        let orderedFlag = getOrderedFlagFrom(contentComponents: contentComponents)
        let entryCount = getEntryCountFrom(contentComponents: contentComponents)
        let childElementIDs = getChildElementIDsFrom(contentComponents: contentComponents)
        let embeddedSubFrames = getEmbeddedSubFrames(contentComponents: contentComponents)
        return ID3FrameTableOfContents(elementID: elementID, topLevelFlag: topLevelFlag, orderedFlag: orderedFlag, entryCount: entryCount, childElementIDs: childElementIDs, embeddedSubFrames: embeddedSubFrames)
    }

    private func getElementIDFrom(contentComponents: [String]) -> String {
        return contentComponents[0]
    }

    private func getTopLevelFlagFrom(contentComponents: [String]) -> Bool {
        let boolInt = Int(contentComponents[1])
        if boolInt == 1 {
            return true
        } else {
            return false
        }
    }

    private func getOrderedFlagFrom(contentComponents: [String]) -> Bool {
        let boolInt = Int(contentComponents[2])
        if boolInt == 1 {
            return true
        } else {
            return false
        }
    }

    private func getEntryCountFrom(contentComponents: [String]) -> Int {
        return Int(contentComponents[3]) ?? 0
    }

    private func getChildElementIDsFrom(contentComponents: [String]) -> [String?] {
        return [contentComponents[4]]
    }

    private func getSubFramesFrom(contentComponents: [String]) -> [ID3Frame] {
        #warning("No idea how to handle this conversion")
        #warning("string to bytes to frame?")
        return contentComponents
    }
}
```

Only it's not going to work, because I don't think there's any way to convert the subframe from a string and back again (even though the subframe is mostly likely just a title, it's still a TIT2 frame.)

So I think we're going to need a new kind of adapter, converting a frame to bytes and back again. I looked through all the documents but if you already have something like that, I don't see it. And I don't have the ability to create it, seeing as how I've done a pretty poor job even trying to create the string adapters for the null-terminated strings of the frames with description strings.

**Comment by chicio** [comment]
Sorry but during these days I don’t have times because I have to follow other stuff at work (and also on another private project). Try to find the best possible solution that keeps the code quality above 3.95 and that works with respect to the id3 standard. If I find some time I would try to go by removing the lyrics tag for the moment from master, so you have much more time to work on this (but given my current personal I doubt I will have spare time to work on this).

**Comment by NCrusher74** [comment]
Well, here's the problem:

I'm working on a project of my own that involves reading and writing audiobook metadata. The sort of informal standard adopted by the audiobook community and manufacturers involves using the comment and/or unsynced lyrics frames for the book blurb/description/summary/whatever. So I really sort of need those fields functional. And the reason I need the language, CTOC and CHAP frames should be obvious.

But I don't have the knowledge yet to implement those frames entirely by myself yet. Your parsing operation/factory and adapter classes don't have much annotation that I can use to understand how they all relate to each other, which means I can't figure out how to emulate them.

And the only other tools that exist that have the functionality I need are C++ libraries like TagLib, and I don't have the skills to create the wrappers I need to use them, either.

Which means I'm stuck

Anyway, thanks for your time. Hopefully you'll find some use for the code I made for your project at some point.

**Comment by NCrusher74** [comment]
Okay, @SDGGiesbrecht has finished his work on the `Comment`, `UnsyncedLyrics` and `UserTextInformation` frames. There's still the preliminary work i did on `Chapter` and `CTOC` on the files, but it's not implemented yet. I think that's for another branch. For now, since this one checks out, maybe you'll want to merge it so that the frames that are already implemented work properly?

**Comment by NCrusher74** [comment]
Hi there. I'm going to close this Pull Request, because I'm in the process of adding another couple dozen frames, including IPLS/TIPL/TMCL frames and Chapter/TableOfContents frames. I'll do another pull request when they're complete.

However...I have a few larger changes I would like to make. With the implementation of Comment/UserDefinedTextInformation/Lyrics frames, `ID3TagEditor` now has frame types for which there can be more than one frame of that type.

In order to get around restrictions about unique keys and values in a `Dictionary`, this is going to require creating a new type (let's call it `FrameKey`, for example) where instead of using `FrameName: ID3Frame` as the `frames` parameter for creating the tag, we would use `FrameKey : ID3Frame` instead. And while `FrameKey` might often be just the `FrameName` for the frames that must be unique, for frames that can be repeated with a different language, description, or whatever, that it might be `FrameKey(description)` or whatever.

@SDGGiesbrecht also has some very good suggestions on why the `FrameName : id3Identifier` dictionaries in `ID3FrameConfiguration` would be better handled as enumerations instead of dictionaries.

But since those changes are a lot more sweeping than just adding new frame types, I wanted to see how you would prefer I handle them, if you'd like me to make them and do a pull request, or just keep them in my own fork, or what.
